components/logins/src/migrate_sqlcipher_db.rs

```diff
@@ -67,27 +101,80 @@ fn sqlcipher_3_compat(conn: &Connection) -> Result<()> {
-        .ok_or_else(|| ErrorKind::InvalidPath(path.to_os_string()))?;
-
-    conn.execute("ATTACH DATABASE ? AS plaintext key ''", &[path])?;
-    // this one is a bit weird because it's a SELECT statement that we know will return 0 rows.
-    // rusqlite will return an Error if we use execute(), so we use query_row with a dummy closure
-    conn.query_row("SELECT sqlcipher_export('plaintext')", NO_PARAMS, |_| Ok(0))?;
-    conn.execute("DETACH DATABASE plaintext", NO_PARAMS)?;
-    Ok(())
+//Manually copy over row by row from sqlcipher db to a plaintext db
+pub fn migrate_from_sqlcipher_db(
+    cipher_conn: &mut Connection,
+    plaintext_store: LoginStore,
+    encryption_key: &str,
+) -> Result<MigrationMetrics> {
+    let start_time = Instant::now();
+
+    let mut metrics: MigrationMetrics = MigrationMetrics::default();
+
+    // Select From both LoginsL and LoginsM with a union to ensure we're covering our
+    // migration cases (one table has but not the other, etc)
+    let mut select_stmt = cipher_conn.prepare(&GET_BY_GUID_SQL)?;
+
+    // Use raw rows to avoid extra copying since we're looping over an entire table
+    let mut rows = select_stmt.query(NO_PARAMS)?;
+    while let Some(row) = rows.next()? {
+        metrics.num_processed += 1;
+        let guid: String = row.get("guid")?;
+        let username: String = row.get("username")?;
+        let password: String = row.get("password")?;
+        // migrating hostname to the new column origin
+        let origin: String = row.get("hostname")?;
+        let http_realm: Option<String> = row.get("httpRealm")?;
+        // migrating formSubmitURL to the new column action origin
+        let form_action_origin: Option<String> = row.get("formSubmitURL")?;
+        let username_field: Option<String> = row.get("usernameField")?;
+        let password_field: Option<String> = row.get("passwordField")?;
+        let time_created: i64 = row.get("timeCreated")?;
+        let time_last_used: i64 = row.get("timeLastUsed").unwrap_or_default();
+        let time_password_changed: i64 = row.get("timePasswordChanged")?;
+        let times_used: i64 = row.get("timesUsed")?;
+
+        // encrypt the username/password data
+        let encryptor = EncryptorDecryptor::new(encryption_key)?;
+        let login: Login = Login {
+            id: guid.to_string(),
+            username_enc: encryptor.encrypt(&username)?,
+            password_enc: encryptor.encrypt(&password)?,
+            origin,
+            http_realm,
+            form_action_origin,
+            username_field: username_field.unwrap_or_default(),
+            password_field: password_field.unwrap_or_default(),
+            // TO_DO: Do we need to convert from microsecond timestamps to milliseconds??
+            time_created,
+            time_last_used,
+            time_password_changed,
+            times_used,
+        };
+
+        // This should be updated with whatever new API we're using to validate, fixup and add
```

Comment:
If the plan is to use `add_or_update` here (there won't be an `add` store function after the API changes) keep in mind that validate and fixup logic will be occurring in that new function. However using that function here may confuse what's happening here. After all we should only ever be using the add logic in `add_or_update`.